binder/src/main/java/io/grpc/binder/internal/BinderClientTransport.java

```diff
@@ -316,14 +316,21 @@ void notifyTerminated() {
       readyTimeoutFuture.cancel(false);
       readyTimeoutFuture = null;
     }
-    if (preAuthResultFuture != null) {
-      preAuthResultFuture.cancel(false); // No effect if already complete.
-    }
-    if (authResultFuture != null) {
-      authResultFuture.cancel(false); // No effect if already complete.
-    }
+
+    ListenableFuture<Status> preAuthFuture = preAuthResultFuture;
+    ListenableFuture<Status> authFuture = authResultFuture;
+    preAuthResultFuture = null;
+    authResultFuture = null;
+
     serviceBinding.unbind();
     clientTransportListener.transportTerminated();
+
+    if (preAuthFuture != null) {
+      offloadExecutor.execute(() -> preAuthFuture.cancel(false)); // No effect if already complete.
+    }
+    if (authFuture != null) {
+      offloadExecutor.execute(() -> authFuture.cancel(false)); // No effect if already complete.
+    }
```

Comment:
The key change is that while the `cancel()` calls are still within the `@GuardedBy("this")` method, they now execute on a different thread via `offloadExecutor.execute()`.